Fabric.Terminology.SqlServer/Persistence/SqlClientTermValueSetRepository.UoWOperation.cs

```diff
@@ -0,0 +1,138 @@
+﻿namespace Fabric.Terminology.SqlServer.Persistence
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    using CallMeMaybe;
+
+    using Fabric.Terminology.Domain;
+    using Fabric.Terminology.Domain.Models;
+    using Fabric.Terminology.Domain.Persistence;
+    using Fabric.Terminology.SqlServer.Models.Dto;
+    using Fabric.Terminology.SqlServer.Persistence.UnitOfWork;
+
+    internal partial class SqlClientTermValueSetRepository
+    {
+        private static ValueSetCodeDto ConvertToValueSetCodeDto(Guid valueSetGuid, ICodeSystemCode code) =>
+            new ValueSetCodeDto(code) { ValueSetGUID = valueSetGuid };
+
+        private AddRemoveCodeOperations PrepareAddRemoveCodes(
+            Guid valueSetGuid,
+            IReadOnlyCollection<ICodeSystemCode> codesToAdd,
+            IReadOnlyCollection<ICodeSystemCode> codesToRemove)
+        {
+            var currentCodeDtos = this.uowManager.GetCodeDtos(valueSetGuid);
+
+            //// There could be duplicates between the codesToAdd and codesToRemove collections.
+            //// In cases where a code is found in both lists, it will simply be removed from each list
+            //// since we are using SqlBulkCopy to batch insert records the "added" codes (e.g. there is no
+            //// update/delete capability in that process and we want to ensure we do not insert duplicate codes)
+
+            // dups are codes that would be removed and then immediately added again.
+            var dups = codesToAdd.Where(code => codesToRemove.Select(r => r.CodeGuid).Contains(code.CodeGuid)).ToList();
+            var codeDeletes = codesToRemove.Except(dups).ToList();
+
+            var batchInsertDtos = codesToAdd.Except(dups).Select(code => ConvertToValueSetCodeDto(valueSetGuid, code)).ToList();
+
+            var removeResult = this.PrepareRemoveCodesOperations(currentCodeDtos, codeDeletes);
+
+            var finalCodes = removeResult.CurrentCodeDtos.Union(batchInsertDtos);
+
+            return new AddRemoveCodeOperations
+            {
+                CurrentCodeDtos = currentCodeDtos,
+                NewCodeDtos = batchInsertDtos,
+                Operations = removeResult.Operations.Union(this.BuildCodeCountOperationList(valueSetGuid, finalCodes))
+                    .ToList()
+            };
+        }
+
+        private IReadOnlyCollection<Operation> BuildRemoveCodesOperationList(
+            IEnumerable<ValueSetCodeDto> originalCodeDtos,
+            IEnumerable<ICodeSystemCode> removeCodes)
+        {
+            var removeGuids = removeCodes.Select(ds => ds.CodeGuid).ToList();
```

Comment:
If you make this a HashSet, it'll reduce the complexity of the following operation.